server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -308,28 +342,142 @@ private void takeSnapshot(Map<String, LookupExtractorFactoryContainer> lookupMap
+      lookupBeanList = getLookupListFromSnapshot();
+    }
+    if (lookupBeanList != null) {
       ImmutableMap.Builder<String, LookupExtractorFactoryContainer> builder = ImmutableMap.builder();
-
-      final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
-      for (LookupBean lookupBean : lookupBeanList) {
-        LookupExtractorFactoryContainer container = lookupBean.getContainer();
-
-        if (container.getLookupExtractorFactory().start()) {
-          builder.put(lookupBean.getName(), container);
-        } else {
-          throw new ISE("Failed to start lookup [%s]:[%s]", lookupBean.getName(), container);
+      ListeningScheduledExecutorService executorService = MoreExecutors.listeningDecorator(
+          Executors.newScheduledThreadPool(
+              lookupConfig.getNumLookupLoadingThreads(),
+              Execs.makeThreadFactory("LookupReferencesManager-Startup-%s")
+          )
+      );
+      try {
+        List<ListenableFuture<Map.Entry>> futures = new ArrayList<>();
+        LOG.info("Starting lookup loading process");
+        for (LookupBean lookupBean : lookupBeanList) {
+          futures.add(
+              executorService.submit(
+                  () -> {
+
+                    LookupExtractorFactoryContainer container = lookupBean.getContainer();
+                    LOG.info(
+                        "Starting lookup [%s]:[%s]",
+                        lookupBean.getName(),
+                        container
+                    );
+                    if (container.getLookupExtractorFactory().start()) {
+                      LOG.info(
+                          "Started lookup [%s]:[%s]",
+                          lookupBean.getName(),
+                          container
+                      );
+                      return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
+                    } else {
+                      LOG.error(
+                          "Failed to start lookup [%s]:[%s]",
+                          lookupBean.getName(),
+                          container
+                      );
+                      return null;
+                    }
+                  }
+              )
+          );
+        }
+        final ListenableFuture<List<Map.Entry>> futureList = Futures.allAsList(futures);
+        try {
+          futureList.get()
+                    .stream()
+                    .filter(Objects::nonNull)
+                    .forEach(builder::put);
+
+          stateRef.set(new LookupUpdateState(builder.build(), ImmutableList.of(), ImmutableList.of()));
         }
+        catch (Exception ex) {
+          futureList.cancel(true);
+          throw ex;
+        }
+      }
+      catch (Exception e) {
+        LOG.error(e, "Failed to finish lookup load process.");
       }
+      finally {
+        try {
+          executorService.shutdownNow();
+          if (!executorService.awaitTermination(60, TimeUnit.SECONDS)) {
```

Comment:
I don't think you really need to wait for termination here, could just call `executorService.shutdownNow();` and exit